File: Android/src/app/src/main/java/com/google/ai/edge/gallery/ui/mainpage/EmotionOverlay.kt
/*
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.ai.edge.gallery.ui.mainpage

import androidx.compose.animation.core.Animatable
import androidx.compose.animation.core.LinearOutSlowInEasing
import androidx.compose.animation.core.tween
import androidx.compose.foundation.layout.BoxWithConstraints
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.key
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.unit.sp
import com.google.ai.edge.gallery.customtasks.voiceassistant.EmotionCue
import kotlin.math.PI
import kotlin.math.sin
import kotlin.random.Random
import kotlinx.coroutines.delay

/**
 * A non-interactive overlay that plays a brief, delightful "emotion" effect — emoji that float up,
 * pop and fade — whenever the assistant's reply contains emoji (instead of reading them aloud). It's
 * meant to make the conversation feel more like an emotional exchange, similar to the floating
 * reactions in modern video-call apps.
 */
@Composable
fun EmotionOverlay(cue: EmotionCue?, modifier: Modifier = Modifier) {
  if (cue == null || cue.emojis.isEmpty()) return
  BoxWithConstraints(modifier = modifier.fillMaxSize()) {
    val widthPx = constraints.maxWidth.toFloat()
    val heightPx = constraints.maxHeight.toFloat()
    // Restart the whole burst when a new cue arrives.
    key(cue.id) { EmotionBurst(emojis = cue.emojis, widthPx = widthPx, heightPx = heightPx) }
  }
}

private data class Particle(
  val emoji: String,
  val startXFraction: Float,
  val driftXPx: Float,
  val swayPx: Float,
  val sizeSp: Int,
  val delayMs: Long,
  val durationMs: Int,
  val rotationDeg: Float,
)

@Composable
private fun EmotionBurst(emojis: List<String>, widthPx: Float, heightPx: Float) {
  val particles =
    remember(emojis, widthPx) {
      val count = 12
      List(count) { i ->
        Particle(
          emoji = emojis[i % emojis.size],
          startXFraction = 0.18f + Random.nextFloat() * 0.64f,
          driftXPx = (Random.nextFloat() - 0.5f) * widthPx * 0.22f,
          swayPx = (12f + Random.nextFloat() * 18f) * (if (Random.nextBoolean()) 1f else -1f),
          sizeSp = (26 + Random.nextInt(22)),
          delayMs = (i * 60 + Random.nextInt(140)).toLong(),
          durationMs = 1500 + Random.nextInt(900),
          rotationDeg = (Random.nextFloat() - 0.5f) * 36f,
        )
      }
    }
  for (p in particles) {
    EmotionParticle(p, widthPx, heightPx)
  }
}

@Composable
private fun EmotionParticle(p: Particle, widthPx: Float, heightPx: Float) {
  val progress = remember { Animatable(0f) }
  LaunchedEffect(Unit) {
    delay(p.delayMs)
    progress.animateTo(1f, animationSpec = tween(durationMillis = p.durationMs, easing = LinearOutSlowInEasing))
  }
  val t = progress.value

  // Fade in quickly, hold, then fade out near the end.
  val alpha =
    when {
      t < 0.12f -> t / 0.12f
      t > 0.7f -> ((1f - t) / 0.3f).coerceIn(0f, 1f)
      else -> 1f
    }
  // Pop in, then settle slightly smaller.
  val scale =
    if (t < 0.25f) lerp(0.5f, 1.15f, t / 0.25f) else lerp(1.15f, 0.92f, (t - 0.25f) / 0.75f)

  // Rise from the lower third toward the upper area, with a gentle horizontal sway.
  val startY = heightPx * 0.72f
  val rise = heightPx * 0.52f
  val y = startY - rise * t
  val x = widthPx * p.startXFraction + p.driftXPx * t + sin(t.toDouble() * PI * 2).toFloat() * p.swayPx

  Text(
    text = p.emoji,
    fontSize = p.sizeSp.sp,
    modifier =
      Modifier.graphicsLayer {
        translationX = x
        translationY = y
        this.alpha = alpha
        scaleX = scale
        scaleY = scale
        rotationZ = p.rotationDeg * t
      },
  )
}

private fun lerp(start: Float, stop: Float, fraction: Float): Float =
  start + (stop - start) * fraction.coerceIn(0f, 1f)

File: iOS/Gallery/UI/MainPage/EmotionOverlay.swift
// Copyright 2025 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Port of ui/mainpage/EmotionOverlay.kt

import SwiftUI
import Foundation

// MARK: - EmotionOverlay

/// A non-interactive overlay that plays a brief "emotion" effect — emoji that float up,
/// pop and fade — whenever the assistant's reply contains emoji (instead of reading them
/// aloud). Mirrors the Android EmotionOverlay composable.
struct EmotionOverlay: View {
    let cue: EmotionCue?

    var body: some View {
        GeometryReader { geo in
            if let cue, !cue.emojis.isEmpty {
                EmotionBurst(emojis: cue.emojis, size: geo.size)
                    .id(cue.id)   // re-trigger when a new cue arrives
            }
        }
        .allowsHitTesting(false)
    }
}

// MARK: - EmotionBurst

private struct ParticleConfig: Identifiable {
    let id: Int
    let emoji: String
    let startXFraction: Double
    let driftXFraction: Double
    let swayPx: Double
    let sizePt: Double
    let delayMs: Double
    let durationMs: Double
    let rotationDeg: Double
}

private struct EmotionBurst: View {
    let emojis: [String]
    let size: CGSize

    @State private var particles: [ParticleConfig] = []

    var body: some View {
        ZStack(alignment: .topLeading) {
            ForEach(particles) { p in
                EmotionParticle(particle: p, size: size)
            }
        }
        .onAppear { buildParticles() }
    }

    private func buildParticles() {
        let count = 12
        particles = (0..<count).map { i in
            ParticleConfig(
                id: i,
                emoji: emojis[i % emojis.count],
                startXFraction: 0.18 + Double.random(in: 0..<0.64),
                driftXFraction: (Double.random(in: 0..<1) - 0.5) * 0.22,
                swayPx: (12 + Double.random(in: 0..<18)) * (Bool.random() ? 1 : -1),
                sizePt: 26 + Double.random(in: 0..<22),
                delayMs: Double(i * 60) + Double.random(in: 0..<140),
                durationMs: 1500 + Double.random(in: 0..<900),
                rotationDeg: (Double.random(in: 0..<1) - 0.5) * 36
            )
        }
    }
}

// MARK: - EmotionParticle

private struct EmotionParticle: View {
    let particle: ParticleConfig
    let size: CGSize

    @State private var progress: Double = 0

    var body: some View {
        let t = progress
        let alpha: Double = {
            if t < 0.12 { return t / 0.12 }
            if t > 0.7  { return max(0, (1 - t) / 0.3) }
            return 1
        }()
        let scale: Double = t < 0.25
            ? lerp(0.5, 1.15, t / 0.25)
            : lerp(1.15, 0.92, (t - 0.25) / 0.75)

        let startY = size.height * 0.72
        let rise   = size.height * 0.52
        let y = startY - rise * t
        let x = size.width * particle.startXFraction
              + size.width * particle.driftXFraction * t
              + sin(t * .pi * 2) * particle.swayPx

        Text(particle.emoji)
            .font(.system(size: particle.sizePt))
            .opacity(alpha)
            .scaleEffect(scale)
            .rotationEffect(.degrees(particle.rotationDeg * t))
            .position(x: x, y: y)
            .onAppear {
                DispatchQueue.main.asyncAfter(deadline: .now() + particle.delayMs / 1000) {
                    withAnimation(.linear(duration: particle.durationMs / 1000)) {
                        progress = 1
                    }
                }
            }
    }
}

private func lerp(_ a: Double, _ b: Double, _ t: Double) -> Double {
    a + (b - a) * min(1, max(0, t))
}
